docs/ldap-config.md

LDAP Configuration


## Steps and instructions to enable LDAP

* Wolf supports LDAP authentication. To enable it, you need to modify the `wolf-server` configuration and then reboot the server.
* After the configuration is complete, the `LDAP` account information will not be synchronized in real-time by `Wolf`. It will only be synced once the user logs into the `Console` or `Agent` of the `Wolf` system.
* If a new user wants to use a Wolf-managed application for the first time, the general steps are:
  * The user logs into the application through the `Agent` and selects the `LDAP` login method.
    * Note: The login may fail due to lack of permissions, but the account information will still be synced to Wolf.
  * The administrator then authorizes the user in the Console.
  * The user can log in again to the application that has been authorized.
* `LDAP` users can only choose to log in using the LDAP method, and not with a password.
* The default `administrator` accounts `root` and `admin` added by Wolf can log in using their account password.

## Configuration

To configure the Wolf system to use LDAP authentication, you need to modify the configuration file `server/conf/config.js`. Add the following configuration (the configuration already has a section for `ldapConfig__` by default, you can change it to `ldapConfig`). Please modify the configuration to match your actual environment:

```js
  ldapConfig: {
    label: 'OpenLDAP',
    url: 'ldap://127.0.0.1:389',
    baseDn: 'dc=example,dc=org',
    adminDn: 'cn=admin,dc=example,dc=org',
    adminPassword: '123456',
    userIdBase: 10000 * 100, // wolf user id = ldap user id + userIdBase
    fieldsMap: { // key=wolf-fieldname, value=ldap-fieldname
      id: 'uidNumber',
      username: 'uid',
      nickname: 'dn',
      email: 'mail',
    },
  },
```

### Configuration Description

The following table provides a description of the configuration keys in the `server/conf/config.js` file for LDAP authentication:

Configuration Key | Description
-------| -------
label | The label that will be displayed as a login option in both `Console` and `Agent`.
url | The URL of the LDAP server.
baseDn | The base DN of the LDAP directory.
adminDn | The DN of the LDAP administrator.
adminPassword | The password of the LDAP administrator.
userIdBase | The base for mapping user IDs when the `LDAP` user is synced to `wolf`. The value of `userIdBase` will be added to the `LDAP` user ID.
fieldsMap | A key-value pair that maps the fields in wolf to the corresponding fields in the LDAP system. The key is the field name in wolf (which can be `id`, `username`, `nickname`, or `email`), and the value is the corresponding field in the LDAP system. The example shows the configuration for `OpenLDAP`.

### Once LDAP authentication is enabled, the login options in both `Console` and `Agent` will automatically include an LDAP login option, as shown in the following screenshot:

![Client Login](./imgs/screenshot/client/login.png)


docs/ldap-config-cn.md

LDAP配置

## 启用LDAP步骤及说明

* wolf支持LDAP认证, 需要修改`wolf-server`配置, 然后重启.
* 配置之后, `wolf`不会主动同步`LDAP`账号信息, 需要用户登录`wolf`系统的`Console`或`Agent`后, 才会同步.
* 新用户首次使用`wolf`管理的应用时, 一般步骤为:
  * 用户通过`Agent`登录应用, 登录方式选择`LDAP`方式.
    * 注意: 登录时, 由于没有相应权限, 会登录失败, 但账号信息会同步到wolf中.
  * 管理员在`Console`中对用户进行授权.
  * 用户再次登录已经被授权的应用.
* `LDAP`用户只能选择使用`LDAP`方式登录, 不能使用账号密码方式.
* `wolf`默认添加的`管理员`账号`root`及`admin`, 可以使用账号密码方式登录.


## 配置

修改配置文件 `server/conf/config.js`, 添加如下配置(配置中默认已经有了`ldapConfig__`, 你可以将它修改为`ldapConfig`):
具体配置请根据实际情况修改.

```js
  ldapConfig: {
    label: 'OpenLDAP',
    url: 'ldap://127.0.0.1:389',
    baseDn: 'dc=example,dc=org',
    adminDn: 'cn=admin,dc=example,dc=org',
    adminPassword: '123456',
    userIdBase: 10000 * 100, // wolf user id = ldap user id + userIdBase
    fieldsMap: { // key=wolf-fieldname, value=ldap-fieldname
      id: 'uidNumber',
      username: 'uid',
      nickname: 'dn',
      email: 'mail',
    },
  },
```

### 配置说明

配置key | 说明
-------| -------
label | 配置的label, 会显示在`Console`与`Agent`的登录选项中.
url | `LDAP`服务器的`url`
baseDn | `LDAP`的基准DN
adminDn | 管理员的DN.
adminPassword | 管理员的密码.
userIdBase | 用户ID映射基数, `LDAP`用户同步到`wolf`时, 会根据`LDAP`用户ID加上 `userIdBase`.
fieldsMap | 字段映射值. 键值对, 键为wolf中的字段名, 可以为`id`,`username`,`nickname`,`email`, 值为ldap系统中对应的字段. 示例中为`OpenLDAP`的配置.

### 启用`LDAP`后, `Console` 及 `Agent` 登录都会自动添加LDAP登录选项, 如下图:

 ![客户端登陆](./imgs/screenshot/client/login.png)



ChangeLog.md
2022-08-01 0.5.5

* `Server` Save the token via md5 key.
* `Server` Remove the RedisCache keys call

2022-07-10 0.5.4

* `Server` Replacing the redis library with ioredis
* `Server` Support for redis as object cache
* `Server` Return 403 http status code when accessing resources without permission

2022-05-25 0.5.3

* `Server`: save login token to redis.
* `Server` & `Console` upgrade dependencies packages.
* `NodeJS` upgraded to v14.17.3
* `npm` upgraded to 6.14.13
* Replaced `node-sass` with `sass`

2021-12-29 0.5.2

* `Console`: selection permission by transfer.

2021-10-22 0.5.0

* Support LDAP authenticate

2021-08-21 0.4.1

* Support MySQL.

2021-07-07 0.4.0

* 服务端口从1008x修改为1218x(解决chrome访问失败的问题).
* `Console`支持中文.

2021-01-14 0.3.3

* 接口添加更严格的参数校验.

2020-05-04 0.3.1

* Application Secret加密保存
* 添加 `WOLF_CRYPT_KEY` 环境变量, 用于设置Secret及OAuth用户ID加密的Key.

2020-05-04 0.3.0

* 登陆支持OAuth2
* API接口被修改为更符合RESTful格式

2020-02-23 0.2.3

* `token` 添加版本
* RBAC 接口添加用户信息获取接口 `/wolf/rbac/user_info`
* 禁用用户
* rbac URL变更
  * `/wolf/rbac/login` ==> `/wolf/rbac/login.html`
  * `/wolf/rbac/login.rest` ==> `/wolf/rbac/login`
  * `/wolf/rbac/change_pwd` ==> `/wolf/rbac/change_pwd.html`
  * `/wolf/rbac/change_pwd.rest` ==> `/wolf/rbac/change_pwd`

2020-02-10 0.2.2

* 日志记录支持查看请求 `body` 或 `args`

2020-01-13 0.2.1

* 添加 `appid` 到 `token` 中.
* url 前缀修改为 `/wolf/`
* 性能优化.日志优化

2020-01-08 0.1.12

* `bcrypt` 模块修改为 `bcrypt-node`

2019-12-22 0.1.11

* 完善单元测试
* 添加 `istanbul/syc` 进行代码覆盖率测试. 
  * 93.95% Statements
  * 78.49% Branches
  * 93.91% Functions 185/197
  * 93.93% Lines 1455/1549
* rbac策略单元测试.

2019-11-16 0.1.10

初始版本.

* RBAC权限控制
* 应用管理
* 用户管理
* 角色管理
* 权限分类
* 权限管理
* 资源管理




server/conf/config.js
const config = {
  rootUserInitialPassword: process.env.RBAC_ROOT_PASSWORD || '123456',
  tokenKey: process.env.RBAC_TOKEN_KEY || 'b5155b92b13a02d08d2cc1bf8b81bec7c0c70fb8',
  cryptKey: process.env.WOLF_CRYPT_KEY || 'fbd4962351924792cb5e5b131435cd30b24e3570',
  rbacTokenExpireTime: parseInt(process.env.RBAC_TOKEN_EXPIRE_TIME) || 3600 * 24 * 30,
  consoleTokenExpireTime: parseInt(process.env.CONSOLE_TOKEN_EXPIRE_TIME) || 3600 * 24 * 30,
  consoleLoginWithCaptcha: ((process.env.CONSOLE_LOGIN_WITH_CAPTCHA || 'no') === 'yes'),
  rbacRecordAccessLog: (process.env.RBAC_RECORD_ACCESS_LOG || 'yes') === 'yes',
  memCacheTTLSecond: 600,
  memCacheByRedis: (process.env.MEM_CACHE_BY_REDIS || 'no') === 'yes',
  clientChangePassword: (process.env.CLIENT_CHANGE_PWD || 'yes') === 'yes',
  database: {
    url: process.env.RBAC_SQL_URL || 'postgres://wolfroot:123456@127.0.0.1:5432/wolf',
  },
  redis: {
    url: process.env.RBAC_REDIS_URL || 'redis://:audit123456@127.0.0.1:6379/0',
    // cluster: [{ port: 6379, host: '127.0.0.1' }],
  },
  oauthOptions: {
    allowEmptyState: false,
    allowExtendedTokenAttributes: true,
    grants: ['authorization_code', 'refresh_token', 'client_credentials', 'password', 'implicit'],
    accessTokenLifetime: parseInt(process.env.OAUTH_ACCESS_TOKEN_LIFETIME) || 3600 * 24 * 7, // 7 days
    refreshTokenLifetime: parseInt(process.env.OAUTH_REFRESH_TOKEN_LIFETIME) || 3600 * 24 * 30, // 30 days.
  },
  ldapConfig__: {
    label: 'OpenLDAP',
    url: 'ldap://127.0.0.1:389',
    baseDn: 'dc=example,dc=org',
    adminDn: 'cn=admin,dc=example,dc=org',
    adminPassword: '123456',
    userIdBase: 10000 * 100, // wolf user id = ldap user id + userIdBase
    fieldsMap: { // key=wolf-fieldname, value=ldap-fieldname
      id: 'uidNumber',
      username: 'uid',
      nickname: 'dn',
      email: 'mail',
    },
  },
}

module.exports = config

// RBAC_SQL_URL = 'mysql://wolfroot:123456@127.0.0.1:3306/wolf'
